api/server/router/image/image_routes.go

```diff
@@ -47,7 +46,11 @@ func (s *imageRouter) postCommit(ctx context.Context, w http.ResponseWriter, r *
 		return err
 	}
 	if c == nil {
-		c = &container.Config{}
+		containerJSON, err := s.backend.ContainerInspectCurrent(cname, false)
```

Comment:
Why is this in builder pkg and not in daemon.commit ?

Reply:
Do you mean the api package? 

`daemon.Commit()` is overloaded right now, it's trying to share too much when the two consumers are not that similar. The parts that are not similar need to be pulled out

Reply:
Sorry, yes, api. Shouldn't `MergeConfigs` mean the same what these lines are doing(when empty config is passed)?

Reply:
My question is why would `MergeConfigs` even need to exist? It seems like a hack to support a weird code structure (it was added in a refactor, not as a feature).

Reply:
My understanding is that commits from builder do not merge while commits from `commit` do.  If you would want to refactor it so that they don't call the same function with different parameters(you comment about it being overloaded) that may be ok, but makes it even weirder to have parts of the same functionality split in different packages.

Reply:
What functionality is split?

Here is my understanding, there are 3 possible operations:
* the builder calls `Commit()` with a new config to save as a child of an image
* the commit endpoint calls `Commit()` with a config that should be applied (merged) over an image config, then saved as a child of the image
* the commit endpoint calls `Commit()` with a list of `changes` that should be translated into a config by the builder, then saved as a child of the image

`Commit()` should only include the parts that are used in these 3 cases (save a config as a child of an image), but right now it does a lot more.

I guess `MergeConfigs` is needed for the second case, but not 1 and 3. 

Does that sound right to you?

> Shouldn't MergeConfigs mean the same what these lines are doing(when empty config is passed)?

It doesn't, because it happens after `BuildFromConfig()`, which is the problem this is trying to fix.

Reply:
> What functionality is split?

Implementation of `MergeConfigs`. It only merges if you specify the override config, not when config is created by changes. To do it properly one has to request the current config before using `daemon.Commit`.

> It doesn't, because it happens after BuildFromConfig()

Exactly, it seemed tp me that `MergeConfigs` is not implemented properly and causes this problem. Suggestion is to fix the problem, not change the caller so that it wouldn't hit this case anymore.

Reply:
Isn't `MergeConfigs` only necessary for case 2 (commit with with config from request body)?

The bug this PR is fixing is in case 3 (commit with `changes` applied by builder). In this case `merge()` is a no-op, because the changes were already applied ("merged") by the builder.

After this change `MergeConfigs` can be removed from `Commit()` and moved out to the caller for the one case that needs it (when `changes` is empty, and request body has a config).

Reply:
This seems to have stalled, is there anything I can do to help this along?

Reply:
@dnephin Unless you change the API to call `dockerfile.BuildFromConfig` I don't see how you could cover all the cases that the API supports.

@asottile Yes, sorry. You can either make the change I asked for or @dnephin can merge because this already has 2 LGTMs.

Reply:
Yes, I think it should call the equivalent of `dockerfile.BuildFromConfig()`. We could add a separate method on `Daemon` if we think we need the abstraction. That is effectively what it does now, except it's mixed up with `Commit()`.

Do you agree with my assessment of the 3 cases?

Reply:
> Do you agree with my assessment of the 3 cases?

2 and 3 could happen together. But that isn't really related to this PR as this is fixing an issue in existing code, not refactoring the internal go api.

Reply:
To be clear, my proposed fixed is 
```
if c.MergeConfigs && c.Config== nil {
  c.Config = container.Config
}
```

before `BuildConfig` in the daemon, that avoids adding the extra method.

Reply:
Thanks. While I still think this makes `Commit()` worse, I agree it does keep the exported interface cleaner. Since it may be a while before we can properly cleanup the commit api backend, I guess this is a better option for now.

@asottile do you mind making that change? It should remove the need to change the Backend interface.

Sorry for all the changes on this.

Reply:
That was actually my initial patch but I believe it broke tests.  I can try it again.

Reply:
seems to work now, probably because of the code I added to avoid the mutation inside `BuildFromConfig`